src/JustEat.StatsD/V2/Buffer.cs

```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Buffers.Text;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace JustEat.StatsD.V2
+{
+    internal ref struct Buffer
+    {
+        public Buffer(Span<byte> source) : this()
+        {
+            Tail = source;
+            Written = 0;
+        }
+
+        public Span<byte> Tail;
+        public int Written;
+    }
+
+    internal static class BufferImpl
```

Comment:
`BufferExtensions` would be a better name given that this class entirely contains extension methods.